validation/validate.ts

```diff
@@ -0,0 +1,329 @@
+  return aomAccNameComputation.result.value;
+}
+
+/**
+ * Gets the AccName for node with [id='idref'] according to Bryan Garaventa's
+ * prototype
+ * @param {CDPSession} client
+ * @param {string} idref
+ */
+async function getBGAccName(
+  client: CDPSession,
+  idref: string
+): Promise<string> {
+  const bgAccNameComputation = (await client.send('Runtime.evaluate', {
+    expression: "getAccName(document.getElementById('" + idref + "')).name;",
+  })) as Protocol.Runtime.EvaluateResponse;
+  return bgAccNameComputation.result.value;
+}
+
+/**
+ * Gets the AccName for node with [id='idref'] according to our implementation
+ * @param {CDPSession} client
+ * @param {string} idref
+ */
+async function getOurAccName(
+  client: CDPSession,
+  idref: string
+): Promise<string> {
+  const ourAccNameComputation = (await client.send('Runtime.evaluate', {
+    expression:
+      "OurLib.getAccessibleName(document.getElementById('" + idref + "'));",
+  })) as Protocol.Runtime.EvaluateResponse;
+  return ourAccNameComputation.result.value;
+}
+
+/*
+
+  Functions / helper functions for minimising the HTML output.
+
+*/
+
+/**
+ * Gets a string of HTML containing only the DOM nodes relevant to the
+ * computation of an accessible name for the node with backendNodeId of
+ * 'nodeId'.
+ * @param {CDPSession} client
+ * @param {DOM.backendNodeId} nodeId
+ */
+async function getMinimisedHTML(
+  client: CDPSession,
+  idref: string
+): Promise<string> {
+  const nodeId = await getBackendNodeIdFromIdref(client, idref);
+  const allNodesInComputation = await getAllNodesInComputation(client, nodeId);
+  const allRelevantNodes = await removeRedundantNodes(
+    client,
+    allNodesInComputation
+  );
+  const minimisedHTML = await getHTMLfromNodeIds(client, allRelevantNodes);
+  return minimisedHTML;
+}
+
+/**
+ * Get all nodes that are connected to nodeId by label references.
+ * The nodes returned by this function are the roots of any
+ * DOM subtrees that are relevant to the computation of nodeId's
+ * accName.
+ * @param {CDPSession} client
+ * @param {DOM.backendNodeId} nodeId
+ */
+async function getAllNodesInComputation(
+  client: CDPSession,
+  nodeId: Protocol.DOM.BackendNodeId
+): Promise<Protocol.DOM.BackendNodeId[]> {
+  const stack: Protocol.DOM.BackendNodeId[] = [];
+  const relatedNodeIds: Protocol.DOM.BackendNodeId[] = [];
+  stack.push(nodeId);
+  // Iterative DFS traverses nodes connected by label references
+  while (stack && stack.length > 0) {
+    const currentNodeId = stack.pop();
+    relatedNodeIds.push(currentNodeId!);
+    const axTree = (await client.send('Accessibility.getPartialAXTree', {
+      backendNodeId: currentNodeId,
+    })) as Protocol.Accessibility.GetPartialAXTreeResponse;
+    // Nodes in axTree.nodes are in order of decreasing depth in the DOM.
+    // We traverse all nodes descendant of currentNodeId, checking whether
+    // each one contains a label reference.
+    for (const axNode of axTree.nodes) {
```

Comment:
`axTree.nodes` contains all descendants, ancestors and children of the current node. Are the relevant ones of these also included in `axNode.name.sources`? If so then I don't think this for loop is necessary, you can just follow `.name.sources`.

Reply:
`name.sources` doesn't seem to provide nodeIds for descendant nodes, so while it does register for a given node that the contents of that node contributed to its name, it doesn't actually allow us to access those nodes (and their outerHTML).

_console.log(node.name?.sources);_
![image](https://user-images.githubusercontent.com/34962541/88660593-f4601780-d0ce-11ea-885d-696be3eacdef.png)

The `name.sources` property is actually a list of `AXValueSource`s, if it contained an extensive list of `AXNodes` it would be _way_ more user friendly in my opinion.